REGRESSION (270199@main): [iOS] Viewport units are incorrect in Safari after entering and exiting fullscreen

pxlcoder opened the pull request:
#### 53929a96966c93804cd8749b573b538503e1665c
<pre>
REGRESSION (270199@main): [iOS] Viewport units are incorrect in Safari after entering and exiting fullscreen
<a href="https://bugs.webkit.org/show_bug.cgi?id=267102">https://bugs.webkit.org/show_bug.cgi?id=267102</a>
<a href="https://rdar.apple.com/120496571">rdar://120496571</a>

Reviewed by Jer Noble.

Following 270199@main, entering and exiting fullscreen results in the overridden
layout parameters (viewLayoutSize, minimumUnobscuredSize, and maximumUnobscuredSize)
getting set to garbage values. This breaks viewport units in Safari, which are
derived from those values.

The intention of 270199@main was to reset these values as if they were never set
while in fullscreen. That change implemented the following methods on `WKWebView`:

```
- (std::optional&lt;CGSize&gt;)_minimumUnobscuredSizeOverride;
- (std::optional&lt;CGSize&gt;)_maximumUnobscuredSizeOverride;
```

However, those methods were already implemented as SPI in another `WKWebView`
category, with different return types:

```
- (CGSize)_minimumUnobscuredSizeOverride;
- (CGSize)_maximumUnobscuredSizeOverride;
```

The end result of implementing the same method twice with different returns
types, is that `std::optional&lt;CGSize&gt;` is treated as `CGSize`, giving garbage
data when these values are attempted to be restored.

Fix by removing the duplicate methods, and refactoring the code to make the
nature of these optionals less confusing. Previously the ivars were optionals,
but the exposed values were CGSize. Now, they are grouped into a single optional
struct.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _recalculateViewportSizesWithMinimumViewportInset:maximumViewportInset:throwOnInvalidInput:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

The key idea behind the refactoring is that `_viewLayoutSizeOverride`,
`_minimumUnobscuredSizeOverride`, and `_maximumUnobscuredSizeOverride` are all
either `std::nullopt` or none of them are `std::nullopt`. This allows for leaving
the underlying type as `CGSize`, with only a single optional struct wrapping
the overridden parameters.

Ultimately, the confusion between methods and ivars with the same name but
different underlying types, which existed before 270199@main, is removed.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):
(-[WKWebView activeViewLayoutSize:]):
(-[WKWebView _frameOrBoundsMayHaveChanged]):
(activeMinimumUnobscuredSize):
(activeMaximumUnobscuredSize):
(-[WKWebView _didStopDeferringGeometryUpdates]):
(-[WKWebView _setAvoidsUnsafeArea:]):
(-[WKWebView _hasOverriddenLayoutParameters]):
(-[WKWebView _minimumLayoutSizeOverride]):
(-[WKWebView _minimumUnobscuredSizeOverride]):
(-[WKWebView _maximumUnobscuredSizeOverride]):
(-[WKWebView _beginAnimatedResizeWithUpdates:]):
(-[WKWebView _overrideLayoutParametersWithMinimumLayoutSize:minimumUnobscuredSizeOverride:maximumUnobscuredSizeOverride:]):
(-[WKWebView _clearOverrideLayoutParameters]):
(-[WKWebView _viewLayoutSizeOverride]): Deleted.
(-[WKWebView _setViewLayoutSizeOverride:]): Deleted.
(-[WKWebView _setMinimumUnobscuredSizeOverride:]): Deleted.
(-[WKWebView _setMaximumUnobscuredSizeOverride:]): Deleted.
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(WebKit::WKWebViewState::applyTo):

The saved values do not need to be optionals, since there is already a check for
`hasOverriddenLayoutParameters`.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/FullscreenOverriddenLayoutParameters.mm: Added.
(swizzledPresentViewController):
(-[FullscreenOverriddenLayoutParametersWebView initWithFrame:configuration:]):
(-[FullscreenOverriddenLayoutParametersWebView enterFullscreen]):
(-[FullscreenOverriddenLayoutParametersWebView exitFullscreen]):
(-[FullscreenOverriddenLayoutParametersWebView didChangeValueForKey:]):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/272679@main">https://commits.webkit.org/272679@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/93f4b154aec3187d615f2eeb81e3186ce9c43d33

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32620 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11373 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34474 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35192 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29486 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33518 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13719 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8560 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28989 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33059 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9583 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29144 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8360 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8492 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29091 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36528 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29645 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29507 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34614 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8615 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6570 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32475 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10296 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7592 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9229 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9216 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -4155,16 +4123,29 @@ - (void)_overrideLayoutParametersWithMinimumLayoutSize:(CGSize)minimumLayoutSize
+
+    overriddenLayoutParameters.viewLayoutSize = CGSizeMake(std::max<CGFloat>(0, minimumLayoutSize.width), std::max<CGFloat>(0, minimumLayoutSize.height));
+
+    ASSERT((!self.bounds.size.width || minimumUnobscuredSizeOverride.width <= self.bounds.size.width) && (!self.bounds.size.height || minimumUnobscuredSizeOverride.height <= self.bounds.size.height));
+    overriddenLayoutParameters.minimumUnobscuredSize = minimumUnobscuredSizeOverride;
+
+    ASSERT((!self.bounds.size.width || maximumUnobscuredSizeOverride.width <= self.bounds.size.width) && (!self.bounds.size.height || maximumUnobscuredSizeOverride.height <= self.bounds.size.height));
+    overriddenLayoutParameters.maximumUnobscuredSize = maximumUnobscuredSizeOverride;
```
Do we need to make these asserts?

pxlcoder replied:
I kept them around from the old setters, and didn't see a compelling reason to remove them.

aprotyas reviewed (commented):
LGTM

pxlcoder committed:
REGRESSION (270199@main): [iOS] Viewport units are incorrect in Safari after entering and exiting fullscreen
https://bugs.webkit.org/show_bug.cgi?id=267102
rdar://120496571

Reviewed by Jer Noble.

Following 270199@main, entering and exiting fullscreen results in the overridden
layout parameters (viewLayoutSize, minimumUnobscuredSize, and maximumUnobscuredSize)
getting set to garbage values. This breaks viewport units in Safari, which are
derived from those values.

The intention of 270199@main was to reset these values as if they were never set
while in fullscreen. That change implemented the following methods on `WKWebView`:

```
- (std::optional<CGSize>)_minimumUnobscuredSizeOverride;
- (std::optional<CGSize>)_maximumUnobscuredSizeOverride;
```

However, those methods were already implemented as SPI in another `WKWebView`
category, with different return types:

```
- (CGSize)_minimumUnobscuredSizeOverride;
- (CGSize)_maximumUnobscuredSizeOverride;
```

The end result of implementing the same method twice with different returns
types, is that `std::optional<CGSize>` is treated as `CGSize`, giving garbage
data when these values are attempted to be restored.

Fix by removing the duplicate methods, and refactoring the code to make the
nature of these optionals less confusing. Previously the ivars were optionals,
but the exposed values were CGSize. Now, they are grouped into a single optional
struct.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _recalculateViewportSizesWithMinimumViewportInset:maximumViewportInset:throwOnInvalidInput:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:

The key idea behind the refactoring is that `_viewLayoutSizeOverride`,
`_minimumUnobscuredSizeOverride`, and `_maximumUnobscuredSizeOverride` are all
either `std::nullopt` or none of them are `std::nullopt`. This allows for leaving
the underlying type as `CGSize`, with only a single optional struct wrapping
the overridden parameters.

Ultimately, the confusion between methods and ivars with the same name but
different underlying types, which existed before 270199@main, is removed.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):
(-[WKWebView activeViewLayoutSize:]):
(-[WKWebView _frameOrBoundsMayHaveChanged]):
(activeMinimumUnobscuredSize):
(activeMaximumUnobscuredSize):
(-[WKWebView _didStopDeferringGeometryUpdates]):
(-[WKWebView _setAvoidsUnsafeArea:]):
(-[WKWebView _hasOverriddenLayoutParameters]):
(-[WKWebView _minimumLayoutSizeOverride]):
(-[WKWebView _minimumUnobscuredSizeOverride]):
(-[WKWebView _maximumUnobscuredSizeOverride]):
(-[WKWebView _beginAnimatedResizeWithUpdates:]):
(-[WKWebView _overrideLayoutParametersWithMinimumLayoutSize:minimumUnobscuredSizeOverride:maximumUnobscuredSizeOverride:]):
(-[WKWebView _clearOverrideLayoutParameters]):
(-[WKWebView _viewLayoutSizeOverride]): Deleted.
(-[WKWebView _setViewLayoutSizeOverride:]): Deleted.
(-[WKWebView _setMinimumUnobscuredSizeOverride:]): Deleted.
(-[WKWebView _setMaximumUnobscuredSizeOverride:]): Deleted.
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
(WebKit::WKWebViewState::applyTo):

The saved values do not need to be optionals, since there is already a check for
`hasOverriddenLayoutParameters`.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/FullscreenOverriddenLayoutParameters.mm: Added.
(swizzledPresentViewController):
(-[FullscreenOverriddenLayoutParametersWebView initWithFrame:configuration:]):
(-[FullscreenOverriddenLayoutParametersWebView enterFullscreen]):
(-[FullscreenOverriddenLayoutParametersWebView exitFullscreen]):
(-[FullscreenOverriddenLayoutParametersWebView didChangeValueForKey:]):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/272679@main
